src/vanilla/utils/atomWithStorage.ts

```diff
@@ -114,6 +129,42 @@ export function createJSONStorage<Value>(
 ): AsyncStorage<Value> | SyncStorage<Value> {
   let lastStr: string | undefined
   let lastValue: Value
+
+  const webStorageSubscribe: Subscribe<Value> = (key, callback) => {
+    if (!(getStringStorage() instanceof window.Storage)) {
+      return () => {}
+    }
+    const storageEventCallback = (e: StorageEvent) => {
+      if (e.storageArea === getStringStorage() && e.key === key) {
+        callback((e.newValue || '') as Value)
+      }
+    }
+    window.addEventListener('storage', storageEventCallback)
+    return () => {
+      window.removeEventListener('storage', storageEventCallback)
+    }
+  }
+
+  const handleSubscribe: SubscribeHandler<Value> = (
+    subscriber,
+    key,
+    callback,
+    initialValue,
+  ) => {
+    function callbackWithParser(v: Value) {
+      let newValue: Value
+      try {
+        newValue = JSON.parse((v as string) || '')
```

Comment:
`v` must be a string always.
```suggestion
        newValue = JSON.parse(v || '')
```

Reply:
No, I mean `v` must be `string` type without any hacks/casts. It feels like there are some lies in types.
I can help for typing later. Please work on other comments first.